src/analysis/processing/qgsalgorithmrectanglesovalsdiamonds.cpp

```diff
@@ -0,0 +1,209 @@
+  if ( mDynamicWidth )
+    mWidthProperty = parameters.value( QStringLiteral( "WIDTH" ) ).value< QgsProperty >();
+
+  mHeight = parameterAsDouble( parameters, QStringLiteral( "HEIGHT" ), context );
+  mDynamicHeight = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "HEIGHT" ) );
+  if ( mDynamicHeight )
+    mHeightProperty = parameters.value( QStringLiteral( "HEIGHT" ) ).value< QgsProperty >();
+
+  mRotation = parameterAsDouble( parameters, QStringLiteral( "ROTATION" ), context );
+  mDynamicRotation = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "ROTATION" ) );
+  if ( mDynamicRotation )
+    mRotationProperty = parameters.value( QStringLiteral( "ROTATION" ) ).value< QgsProperty >();
+
+  mSegments = parameterAsDouble( parameters, QStringLiteral( "SEGMENTS" ), context );
+
+  return true;
+}
+
+QgsFeatureList QgsRectanglesOvalsDiamondsAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsFeature outFeature = feature;
+  if ( outFeature.hasGeometry() )
+  {
+    double width = mWidth;
+    if ( mDynamicWidth )
+      width = mWidthProperty.valueAsDouble( context.expressionContext(), width );
+
+    double height = mHeight;
+    if ( mDynamicHeight )
+      height = mHeightProperty.valueAsDouble( context.expressionContext(), height );
+
+    double rotation = mRotation;
+    if ( mDynamicRotation )
+      rotation = mRotationProperty.valueAsDouble( context.expressionContext(), rotation );
+
+    if ( width == 0 || height == 0 )
+    {
+      throw QgsProcessingException( QObject::tr( "Width and height should be greater than 0." ) );
+    }
+
+    double phi = rotation * M_PI / 180;
+    double xOffset = width / 2.0;
+    double yOffset = height / 2.0;
+    QgsPointXY point = outFeature.geometry().asPoint();
+    double x = point.x();
+    double y = point.y();
+
+    QgsPolylineXY ring;
+    switch ( mShape )
+    {
+      case 0:
+        // rectangle
+        ring << QgsPointXY( -xOffset, -yOffset )
+             << QgsPointXY( -xOffset, yOffset )
+             << QgsPointXY( xOffset, yOffset )
+             << QgsPointXY( xOffset, -yOffset );
+        break;
+      case 1:
+        // diamond
+        ring << QgsPointXY( 0.0, -yOffset )
+             << QgsPointXY( -xOffset, 0.0 )
+             << QgsPointXY( 0.0, yOffset )
+             << QgsPointXY( xOffset, 0.0 );
+        break;
+      case 2:
+        // oval
+        for ( int i = 0; i < mSegments; i ++ )
+        {
+          double t = ( 2 * M_PI ) / mSegments * i;
+          ring << QgsPointXY( xOffset * cos( t ), yOffset * sin( t ) );
+        }
+        break;
+    }
+
+    for ( int i = 0; i < ring.size(); ++i )
+    {
+      QgsPointXY p = ring[ i ];
+      double px = p.x();
+      double py = p.y();
+      ring[ i ] = QgsPointXY( px * cos( phi ) + py * sin( phi ) + x,
+                              -px * sin( phi ) + py * cos( phi ) + y );
+    }
+
```

Comment:
If input point has z or m, call addZValue or addMValue on the created geometry too

Reply:
Done